### PR TITLE
[ast & semantics] Make more type information available

### PR DESCRIPTION
**Pull request by systay**
## Description
The new evalengine [compiler](https://github.com/vitessio/vitess/pull/12369) needs full type information to be as efficient as possible.  This PR updates bindvars and offsets to give the compiler more type information.

### Bindvars

We usually auto-parameterize (normalize) queries by extracting literals into parameters (bindvars). This makes it possible for the multiple invocations of the same query, with different literal values, to share a single query plan. When we parameterize literals, we are also extracting valuable type information from the query.

This PR tries to address this by adding type information to the bindvars. Example:

```sql
select * from tbl id = 42
```

This would get normalized to:

```sql
select * from tbl id = :id
```

With the changes in this PR, we will instead normalize literals to include their type with a comment:

```sql
select * from tbl id = :id /* INT64 */
```

### Offsets

In this PR, I've also changed `Offset`s to include the original expression, which gives the evalengine compiler a chance to include type information when compiling offsets.

## Related Issue(s)
Supports https://github.com/vitessio/vitess/pull/12369

### CODE REVIEW

File: go/vt/sqlparser/ast_rewriting.go [hunk 1]
```diff
@@ -137,6 +139,15 @@ func (r *ReservedVars) ReserveHasValuesSubQuery() string {
...
 
+func typePostFix(typ sqltypes.Type) string {
+	if typ == sqltypes.Tuple || typ == sqltypes.Null {
+		// tuples already have specific syntax, so we don't need to add anything here
+		// null signals that we don't have type information here
+		return ""
+	}
+	return "." + typ.String()
+}
+
```
Comment (vmg) [inline go/vt/sqlparser/ast_rewriting.go:150]: this function is unused now

File: go/vt/sqlparser/normalizer.go [hunk 2]
```diff
@@ -331,6 +329,33 @@ func (nz *normalizer) convertUpdateExpr(node *UpdateExpr) {
...
+	case BitVal:
+		return sqltypes.HexNum
+	case DateVal:
+		return sqltypes.Date
+	case TimeVal:
+		return sqltypes.Time
+	case TimestampVal:
+		return sqltypes.Datetime
+	default:
+		return 0
```
Comment (systay) [inline go/vt/sqlparser/normalizer.go:355]: @vmg should this be -1 instead?

**Review by GuptaManan100 — COMMENTED** [review]
I like this way of adding the type in comments instead of using a delimiter. Definitely cleaner. However, this approach requires changes in parsing too. MySQL doesn't allow comments to be parsed anywhere. We have to change the parser to accept comments after a bindVariable, otherwise they get ignored. For example, I added a parse test - 
```go
=== RUN   TestOne
    parse_test.go:5407: 
        	Error Trace:	/Users/manangupta/vitess/go/vt/sqlparser/parse_test.go:5407
        	Error:      	Not equal: 
        	            	expected: "select comms_by_companies.* from comms_by_companies where comms_by_companies.id = :comms_by_companies_id /* VARCHAR */ limit :bv1 /* INT64 */, :bv2 /* INT64 */"
        	            	actual  : "select comms_by_companies.* from comms_by_companies where comms_by_companies.id = :comms_by_companies_id limit :bv1, :bv2"
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -1 +1 @@
        	            	-select comms_by_companies.* from comms_by_companies where comms_by_companies.id = :comms_by_companies_id /* VARCHAR */ limit :bv1 /* INT64 */, :bv2 /* INT64 */
        	            	+select comms_by_companies.* from comms_by_companies where comms_by_companies.id = :comms_by_companies_id limit :bv1, :bv2
        	Test:       	TestOne
--- FAIL: TestOne (0.00s)
```

**Comment by GuptaManan100** [comment]
Discussed with @vmg  and we don't really need parser changes since we don't need this type information on the vttablets. On the vtgates, we'll be using the AST everywhere, so we won't need this functionality.

**Review by GuptaManan100 — APPROVED** [review]

**Pull request merged**